components/properties/src/provider.rs

```diff
@@ -311,11 +314,128 @@ impl<'data, T: TrieValue> PropertyCodePointMapV1<'data, T> {
+/// https://www.unicode.org/Public/UCD/latest/ucd/PropertyValueAliases.txt
+///
+/// (matched case-insensitively in ASCII, ignoring underscores, whitespace, and hyphens)
+///
+/// This is expected to be ASCII, but we do not rely on this invariant anywhere except during
+/// datagen.
+#[derive(PartialEq, Eq)] // VarULE wants these to be byte equality
+#[zerovec::make_varule(NormalizedPropertyName)]
+#[cfg_attr(feature = "datagen", zerovec::derive(Serialize))]
+#[cfg_attr(feature = "serde", zerovec::derive(Deserialize))]
+#[zerovec::skip_derive(Ord)]
+pub struct NormalizedPropertyNameBorrowed<'a>(&'a [u8]);
+
+#[cfg(feature = "datagen")]
+impl<'a> serde::Serialize for NormalizedPropertyNameBorrowed<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::Error;
+        if serializer.is_human_readable() {
+            let s = str::from_utf8(self.0)
+                .map_err(|_| S::Error::custom("Attempted to datagen invalid string property"))?;
+            s.serialize(serializer)
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de: 'a, 'a> serde::Deserialize<'de> for NormalizedPropertyNameBorrowed<'a> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            <&str>::deserialize(deserializer).map(|s| Self(s.as_bytes()))
+        } else {
+            <&[u8]>::deserialize(deserializer).map(Self)
+        }
+    }
+}
+
+impl PartialOrd for NormalizedPropertyName {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Normalize a character based on the "loose matching" described in PropertyValueAliases.txt,
+/// returning None for skippable characters
+///
+/// ICU has [code for this][1] (and [during property lookup][2]) which we emulate.
+/// In particular, ICU only does normalization within ASCII, which makes sense since character names
+/// seem to be only ASCII.
+///
+/// [1]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L35
+/// [2]: https://github.com/unicode-org/icu/blob/288c4c7555915ce7b1fb675d94ddd495058fc039/icu4c/source/common/propname.cpp#L226-L230
+fn normalize_char(ch: u8) -> Option<u8> {
+    match ch {
+        b'a'..=b'z' => Some(ch),
+        b'A'..=b'Z' => Some(ch | 0x20), // Uppercase by setting last bit
+        // underscores, hyphens, and all ASCII whitespace
+        b'_' | b'-' | b' ' | 0x09..=0x0d => None,
+        _ => Some(ch),
+    }
+}
+
+impl Ord for NormalizedPropertyName {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_iter = self.0.iter().copied().filter_map(normalize_char);
+        let other_iter = other.0.iter().copied().filter_map(normalize_char);
+        self_iter.cmp(other_iter)
```

Comment:
Praise: This is a really clean Ord impl

Reply:
I was bracing myself for something gnarly and then realized I could do this :smile: